lib/crawlers/base_crawler.rb

```diff
@@ -0,0 +1,153 @@
+    # @return [Array]
+    attr_reader :hits
+
+    # Crawl limiters
+
+    # @return [Integer, NilClass]
+    attr_accessor :max_crawl
+    # @return [Float, NilClass]
+    attr_accessor :max_time
+    # @return [Integer, NilClass]
+    attr_accessor :max_hits
+
+    # @param url [String]
+    #
+    # @param stubs_path [Pathname, NilClass]
+    def initialize(url, stubs_path = nil)
+      @starting_url = url
+      @hits = []
+      if stubs_path.nil?
+        puts "No priority stub path given."
+        @priority_stubs = []
+      else
+        @priority_stubs = load_stubs(stubs_path)
+      end
+    end
+
+    private
+
+    # Load priority stubs to use.
+    # Expects file format to be plain text, one stub per line
+    #
+    # @param path [Pathname]
+    #
+    # @return [Array<String>]
+    def load_stubs(path)
+      return path.readlines.map(&:chomp)
+    end
+
+    # Get current timestamp
+    #
+    # @return [Float]
+    def current_time
+      return Process.clock_gettime(Process::CLOCK_MONOTONIC)
+    end
+
+    # Return elapsed time
+    #
+    # @return [Float]
+    def elapsed_time
+      return current_time - @start_time
+    end
+
+    # Return `true` if max crawl exceeded
+    #
+    # @param agent [Spidr::PriorityAgent]
+    #
+    # @return [TrueClass, FalseClass]
+    def max_crawl_exceeded?(agent)
+      return false unless !@max_crawl.nil? && @max_crawl < agent.history.length
+
+      return true
+    end
+
+    # Return `true` if max time exceeded
+    #
+    # @return [TrueClass, FalseClass]
+    def max_time_exceeded?
+      return false unless !@max_time.nil? && @max_time <= elapsed_time
+
+      return true
+    end
+
+    # Return `true` if max hits exceeded
+    #
+    # @return [TrueClass, FalseClass]
+    def max_hits_exceeded?
+      return false unless !@max_hits.nil? && @hits.length >= @max_hits
+
+      return true
+    end
+
+    # Return `true` if any limit is exceeded
+    #
+    # @param agent [Spidr::PriorityAgent]
+    #
+    # @return [TrueClass, FalseClass]
+    def limits_exceeded?(agent)
+      return max_crawl_exceeded?(agent) || max_time_exceeded? || max_hits_exceeded?
+    end
+
+    # Return the domain from a url, stripping `www.` if present
+    # i.e. https://www.website.com -> website.com
+    #
+    # @param url [String]
+    #
+    # @return [String]
+    def url_to_domain(url)
+      return URI(url).host.delete_prefix('www.')
+    end
+
+    # Override to implement per page behavior
+    #
+    # @param page [Spidr::Page]
+    def process_page(page)
+      raise NotImplementedError
+    end
+
+    public
+
+    # Start crawling with the specified limits, if any.
+    # Returns the contents of `@hits`.
+    #
+    # @param max_crawl [Integer, NilClass]
+    #
+    # @param max_time [Float, NilClass]
+    #
+    # @param max_hits [Integer, NilClass]
+    #
+    # @return [Array]
+    def crawl(max_crawl = nil, max_time = nil, max_hits = nil)
```

Comment:
Should this method be separated out? Ideally methods have max 5 lines as per sandi metz (not strictly obviously)

Reply:
Got it down to 4 if you don't count 'end'